pkg/deployment/cert.go

```diff
@@ -0,0 +1,163 @@
+)
+
+// GenerateTLSCerts generates secrets containing certificates for relevant services
+func GenerateTLSCerts(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1.OpenStackDataPlaneNodeSet,
+	allIPSets map[string]infranetworkv1.IPSet) error {
+
+	// We need to create a secret for each service that contains all the certs needed for
+	// that service for all the nodes in the nodeset.  This one secret will be mounted in
+	// the ansibleEE pod as an extra mount when that service is deployed.
+	//
+	// The convention we use here is "<nodeset.name>-<service>-certs", so for example,
+	// openstack-epdm-nova-certs.
+	novaCertsData := map[string][]byte{}
+	libvirtCertsData := map[string][]byte{}
+
+	// for each node in the nodeset, issue all the TLS certs needed based on the
+	// ips or DNS Names
+	for nodeName, _ := range instance.Spec.Nodes {
+		var ips []string
+		var dnsNames []string
+		var secretName string
+		var certName string
+
+		// TODO(alee) decide if we want to use other labels
+		// For now we just add the hostname so we can select all the certs on one node
+		labels := map[string]string{
+			"hostname": nodeName,
+		}
+
+		ipSet, ok := allIPSets[nodeName]
+		if ok {
+			for _, res := range ipSet.Status.Reservation {
+				ips = append(ips, res.Address)
+				fqdnName := strings.Join([]string{nodeName, res.DNSDomain}, ".")
+				dnsNames = append(dnsNames, fqdnName)
+			}
+		}
+
+		// for now, just get a single cert with all the dnsnames for the host.  Its likely
+		// that this will be sufficient for most services.
+		// But it is clear how we can issue different certs as needed by calling IssueTLSCert
+		// and using ips instead of dnsnames, or by using a different issuer
+		secretName = "cert-" + nodeName
+		certName = secretName
+		certSecret, err := IssueTLSCert(ctx, helper, certName, instance.Namespace, secretName,
+			nodeName, dnsNames, nil, RootCAIssuerInternalNamespaceLabel,
+			RootCAIssuerInternalNamespaceLabel, labels)
+		if err != nil {
+			return err
+		}
+
+		// TODO(alee) Add an owner reference to the secret so it can be monitored
+		// We'll do this once stuggi adds a function to do this in libcommon
+
+		// To use this cert, add it to the relevant service data
+		for key, value := range certSecret.Data {
+			novaCertsData[nodeName+"-"+key] = value
+			libvirtCertsData[nodeName+"-"+key] = value
+		}
+	}
+
+	// create a secret for each service
+	novaCertsSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: instance.Name + "-nova-certs",
+		},
+		Data: novaCertsData,
+	}
+	secret.CreateOrPatchSecret(ctx, helper, instance, novaCertsSecret)
+
+	libvirtCertsSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: instance.Name + "-libvirt-certs",
+		},
+		Data: libvirtCertsData,
+	}
+	secret.CreateOrPatchSecret(ctx, helper, instance, libvirtCertsSecret)
+
+	return nil
+}
+
+func IssueTLSCert(ctx context.Context, helper *helper.Helper,
```

Comment:
> I think functions like this should go in `lib-common` so that we're able to generate certs in openstack-operator 

You have a good point about `lib-common` but I'm not sure that `openstack-operator` should be creating the certs for the EDPM nodes.

When would the openstack-operator know of the DNS names for the EDPM nodes? I think Ade is creating the certs in the dataplane-operator because when node sets are created, that is the moment the DNS names are known, and thus the certs can be created. I'm not sure it could be earlier in the deployment process.

Now, there is certificate related code in lib-common that the dataplane-operator could call.

  https://github.com/openstack-k8s-operators/lib-common/tree/main/modules/certmanager

Both this `cert.go` and the above `certificate.go` in lib-common use `github.com/cert-manager` so maybe certificate.go in lib-common should be used instead to create certificates and what's in the dataplane operator should be calling that code. I haven't looked at it closely  but  both seem to have certificate creating functions.

So I'm fine with the calls to generate the certs for edpm nodes, being made in the dataplane operator, though I agree it's worth considering using lib-common.

Reply:
not saying that `openstack-operator` should create certs for the `edpm` nodes, sorry if I didn't write it properly. My only point is that some logic (e.g., cert generation) is required in the ctlplane, where you want to Create certs and use them to secure the generated Routes [1]. The same certs are required to re-encrypt the traffic from the Route to the service operator, hence the openstack operator should provide them to the service operators.
In terms of edm nodes creation/generation, I agree that the entire process should be driven by the `dataplane-operator`, that could call the same functions provided by `lib-common` with different input coming from the node sets definition. 

[1] https://docs.openshift.com/dedicated/networking/routes/secured-routes.html

Reply:
agreed.  this function should likely go into lib-common.  I'll discuss further with @stuggi

Reply:
we already have one in lib-common https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/certmanager/certificate.go#L137

Reply:
Great.  I'll use that one.
That said  - when anyone (maybe ovn?) needs it - we'll need to add support for IPs instead of just hostnames

Reply:
Latest code uses EnsureCerts in lib-common